### PR TITLE
[hist] Avoid `try-catch` in TH1

### PR DESCRIPTION
**Pull request by amecca**
# This Pull request:

## Changes or fixes:
Made the static function `TH1::CheckConsistency` public and changed the return type to `Bool_t`.
The exceptions types it throws were also made public, so that user code could catch them

## Checklist:

- [X] tested changes locally

## Motivation
At the moment `TH1::CheckConsistency` is called internally by `TH1::Add`, and prints error messages in case of failure, but continues the execution.
Making it public allows user code to manually check the consistency to catch errors and stop the program.
This is especially useful for programs that make hundreds of `Add`s, out of which only a few fail, and the error messages may be lost in logs and go unnoticed.

### CODE REVIEW

**Comment by vepadulano** [comment]
@phsft-bot build

**Comment by vepadulano** [comment]
Thank you @amecca for opening this PR! While we wait for a full review, can I ask you what is the reasoning behind changing the return value type from `bool` to `Bool_t`?

**Comment by amecca** [comment]
Thanks for your prompt reply @vepadulano.
I changed the return type because I thought that it was preferred to use types defined in `Rtypes.h` in public interfaces.
Now that I look more in detail at the coding convention recommendations I see that this only applies to data that mught get written to disk.
I can revert it; do you prefer that I make another commit on top or should I squash them?

**Comment by vepadulano** [comment]
>  I can revert it; do you prefer that I make another commit on top or should I squash them?

No problem, it's up to you for now. Most probably, after the PR review and before merging it, all the smaller commits will be squashed into one to avoid cluttering the commit history too much

**Comment by amecca** [comment]
@vepadulano, @lmoneta, sorry for the ping.
I see that some of the tests fail, but as far as I can tell the failures are not related to this PR.
Do you want me to do something, such as rebase the branch onto a more stable commit (which passes the tests) and try again?

**Comment by pcanal** [comment]
@phsft-bot build

**Review by guitargeek — APPROVED** [review]
Hi @amecca, thanks for the initiative!

Before making this `CheckConsistency()` public, I had to change the interface because so far it uses exceptions to manage the control flow, which is pretty bad practice. Exceptions should not be used to deal with expected cases like this.

I have an alternative suggestion with a return value that encodes the inconsistency. What do you think?

I would be happy if you could test this and give feedback! (given that the CI passes first)

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
> Do you want me to do something, such as rebase the branch onto a more stable commit (which passes the tests) and try again?

Just FYI: the PR is always automatically rebased on top of `master` for the test, so we have no control over that. We just need to keep ROOT `master` stable :)

**Comment by guitargeek** [comment]
The test failures are not serious, as far as I know. The new enum values just need to be added to some reference file in `roottest`.

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
Hi! Since I didn't get feedback yet, I decided to just merge the refactor to check the consistency without the try-catch hack.

If you need the public interface after all, fee free to open a new PR about that once this one is merged :+1:

**Pull request merged**
